### PR TITLE
use `confstr(_CS_DARWIN_USER_TEMP_DIR, ...)` as a `TMPDIR` fallback on Darwin

### LINKED ISSUE #99608: Consider returning `/private/tmp` instead of `/tmp` on macOS with `std::env::temp_dir()`
The first is a regular directory, the second is a symlink to the first. This can lead to software failing in strange and unexpected ways (especially during testing, with the tempfile crate or similar) if symlinks are forbidden, e.g. I recently encountered the case with SQLite

Since `/tmp -> /private/tmp` software that was already working correctly should not start to fail unexpectedly. Software that *was* failing/had to workaround this will start working correctly.

```
macos-12.4 $ /bin/ls -la /tmp
lrwxr-xr-x@ 1 root  wheel  11  9 mai 23:30 /tmp -> private/tmp
```

### PR DESCRIPTION
**Pull request by thomcc**
Followup to #100196 ~~but should wait on https://github.com/rust-lang/libc/pull/2883 before anything is done with it.~~ ~~(Edit: Now it needs https://github.com/rust-lang/libc/pull/2931... *sigh*)~~

See https://github.com/rust-lang/rust/pull/100824#issuecomment-1262264127 for an explanation of the change this makes and why I think this is worth doing.

Closes #99608.

### CODE REVIEW

**Comment by thomcc** [comment]
This is a behavioral change so should go through FCP probably.

See https://github.com/rust-lang/rust/pull/100196#issuecomment-1207449201 and https://github.com/rust-lang/rust/pull/100196#issuecomment-1220067550.

I'll also clean up the `#[cfg(target_vendor = "apple")]` to be the OSes in question, when I make the change following the API being available in libc.

**Comment by thomcc** [comment]
Now it needs https://github.com/rust-lang/libc/pull/2931... *sigh*.

**Comment by thomcc** [comment]
This is a behavioral change in an edge case on Darwin platforms (macOS, iOS, ...).

Specifically, this changes it so that iff `TMPDIR` is unset in the environment, then we use `confstr(_CS_DARWIN_USER_TEMP_DIR, ...)` to query the user temporary directory (previously we just returned `"/tmp"`). If this fails (probably possible in a sandboxed program), only then do we fallback to `"/tmp"` (as before).

The motivation here two-fold:
1. This is in line with the [platform security recommendations](https://developer.apple.com/library/archive/documentation/Security/Conceptual/SecureCodingGuide/Articles/RaceConditions.html#//apple_ref/doc/uid/TP40002585-SW10), as it is unavailable to other users (although it is the same value as seen by all other processes run by the same user).
2. This is a more consistent fallback for when `getenv("TMPDIR")` is unavailable, as `$TMPDIR` is usually initialized to the `DARWIN_USER_TEMP_DIR`.

It seems quite unlikely that anybody will break because of this, and I think it falls under the carve-out we have for platform specific behavior: https://doc.rust-lang.org/nightly/std/io/index.html#platform-specific-behavior.

**Comment by edmorley** [comment]
@thomcc Hi! We would love to see this land - is this in an FCP queue, or does it need to be manually submitted to that process?

**Comment by dtolnay** [comment]
@rust-lang/libs-api:
@rfcbot fcp merge

Relevant reading:

1. https://github.com/rust-lang/rust/pull/100196#issuecomment-1207449201
2. https://github.com/rust-lang/rust/pull/100196#issuecomment-1207440442
3. https://github.com/rust-lang/rust/issues/99608#issue-1315107148

Link 2 is compelling as to why this PR is the best solution. Link 1 lists our other options, neither of which is better.

**Review by dtolnay — CHANGES_REQUESTED** [review]
> I'll also clean up the `#[cfg(target_vendor = "apple")]` to be the OSes in question and move it out of a draft PR, when I make the change following the API being available in libc.

Waiting for this (quoted from https://github.com/rust-lang/rust/pull/100824#issuecomment-1221426001).

LGTM as soon as the cfg is fixed.

**Comment by Dylan-DPC** [comment]
@thomcc any updates on this? thanks

**Review by madsmtm — COMMENTED** [review]
Bump @thomcc?

I can rebase, fix the remaining `#[cfg(target_vendor = "...")]` and submit an updated PR if you want?

@rustbot label O-apple

**Comment by Dylan-DPC** [comment]
@madsmtm yes feel free to do that :)

File: library/std/src/sys/unix/os.rs [hunk 1]
```diff
@@ -579,12 +579,80 @@ pub fn page_size() -> usize {
...
+        buf.reserve(bytes_needed_including_nul);
+        // `confstr` returns
+        // - 0 in the case of errors: we break and return an error.
+        // - The number of bytes written, iff the provided buffer is enough to
+        //   hold the entire value: we break and return the data in `buf`.
+        // - Otherwise, the number of bytes needed (including nul): we go
+        //   through the loop again.
+        bytes_needed_including_nul =
+            unsafe { libc::confstr(key, buf.as_mut_ptr().cast::<c_char>(), buf.capacity()) };
+    }
```
Comment (madsmtm) [inline library/std/src/sys/unix/os.rs:619]: It is not entirely clear to me why we do all this work, it seems like just using a static `[u8; PATH_MAX]` would be sufficient, and in any case, we can always just fall back to `/tmp` if not.

But regardless, this works, and I guess there's not much harm in it, so I won't belabour the point.

**Review by madsmtm — COMMENTED** [review]
I've submitted https://github.com/rust-lang/rust/pull/131505

File: library/std/src/sys/unix/os.rs (hunk 1 above)
Comment (BlackHoleFox) [reply library/std/src/sys/unix/os.rs:619]: Unsure about the loop but the retry is also done by Swift's implementation of this: https://github.com/swiftlang/swift-corelibs-foundation/blob/2d23cf3dc07951ed2b988608d08d7a54cc53b26e/Sources/Foundation/NSPathUtilities.swift#L38

**Pull request closed without merging**